pkg/gardenlet/controller/federatedseed/extensions/controllerinstallation_control.go

```diff
@@ -0,0 +1,216 @@
+				if resource.Kind == kind {
+					controllerRegistrationNamesForKind.Insert(controllerRegistration.Name)
+					break
+				}
+			}
+		}
+
+		// Step 3: List all existing controller installation objects for the seed cluster this controller is responsible
+		// for and filter for those that reference registrations collected above. Then requeue those installations for
+		// the other reconciler to decide whether it is required or not.
+
+		controllerInstallationList := &gardencorev1beta1.ControllerInstallationList{}
+		if err := c.k8sGardenClient.Client().List(ctx, controllerInstallationList); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		for _, obj := range controllerInstallationList.Items {
+			if obj.Spec.SeedRef.Name != c.seedName {
+				continue
+			}
+
+			if !controllerRegistrationNamesForKind.Has(obj.Spec.RegistrationRef.Name) {
+				continue
+			}
+
+			c.controllerInstallationQueue.Add(obj.Name)
+		}
+
+		return reconcile.Result{}, nil
+	}
+}
+
+// createControllerInstallationRequiredReconcileFunc reconciles ControllerInstallations. It checks whether it is still
+// required by using the <kindToRequiredTypes> map that was computed by a separate reconciler.
+func (c *controllerInstallationControl) createControllerInstallationRequiredReconcileFunc(ctx context.Context) reconcile.Func {
+	return func(req reconcile.Request) (reconcile.Result, error) {
+		controllerInstallation := &gardencorev1beta1.ControllerInstallation{}
+		if err := c.k8sGardenClient.Client().Get(ctx, req.NamespacedName, controllerInstallation); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		controllerRegistration := &gardencorev1beta1.ControllerRegistration{}
+		if err := c.k8sGardenClient.Client().Get(ctx, kutil.Key(controllerInstallation.Spec.RegistrationRef.Name), controllerRegistration); err != nil {
+			return reconcile.Result{}, err
+		}
+
+		var (
+			allKindsCalculated = true
+			required           *bool
+			requiredKindTypes  = sets.NewString()
+			message            string
+		)
+
+		c.lock.RLock()
+		for _, resource := range controllerRegistration.Spec.Resources {
+			requiredTypes, ok := c.kindToRequiredTypes[resource.Kind]
+			if !ok {
+				allKindsCalculated = false
+				continue
+			}
+
+			if requiredTypes.Has(resource.Type) {
+				required = pointer.BoolPtr(true)
+				requiredKindTypes.Insert(fmt.Sprintf("%s/%s", resource.Kind, resource.Type))
+			}
+		}
+		c.lock.RUnlock()
+
+		if required == nil {
+			if !allKindsCalculated {
+				// if required wasn't set yet then but not all kinds were calculated then the it's not possible to
+				// decide yet whether it's required or not
+				return reconcile.Result{}, nil
+			}
+
+			// if required wasn't set yet then but all kinds were calculated then the installation is no longer required
+			required = pointer.BoolPtr(false)
+			message = "no extension objects exist in the seed having the kind/type combinations the controller is responsible for"
```

Comment:
```suggestion
			message = "No extension objects exist in the seed having the kind/type combinations the controller is responsible for."
```

Reply:
Kubelet doesn't use this style either and I'd keep the message style similar to [Golang error strings](https://github.com/golang/go/wiki/CodeReviewComments#error-strings)